# Assignment

## Group Members

- Kaushik Kumaran _(kk34537)_
- Rohitashwa Chakraborty _(rc47878)_
- Sahil Arora _(sa52623)_


# Question-1: Rock, Paper, Scissors (20 points)

## Game setup:

- Each game round consists two turns, the first by the computer and the second by a human.

- The computer continues playing rounds untill the human chooses to quit.

- The computer chooses one of Rock, Paper, and Scissors, but keeps its choice secret.

- The computer asks for the human’s input.

- The human chooses one of Rock, Paper, and Scissors, or Quit.

- Unless the human quits, the computer figures out the result of the game, as follows:
  - Rock smashes Scissors, so Rock beats Scissors
  - Scissors can cut up paper, so Scissors beat Paper
  - Paper covers Rock, so Paper beats Rock
  - If both players chose the same, it is a draw
  - The computer reports theresult of this round

- If the human chooses to quit, the computer reports:
  - Number of games played
  - Number of times the human won


## Importing the relevant liabraries

In [37]:
## Loading liabraries to work
import numpy as np
## Command to display all outputs in Jupyter 
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Defining constant parameters that we will use in the code later 

Choice List: A list of possible choices that can be made

winning dictionary : This is the dictionary to choose the weapon against the Person so that computer can win the game

In [38]:

### Lets first associate a dictionary to the possible outcomes : R: R; P:P ; S:S ; Quit:Q
choice_list = ['R','P','S','Q']


## Defining Winning Dictionary
winning_dictionary = {
    'S':'R',
    'P':'S',
    'R':'P'}

## Functions for computer choice

most_frequent:

    1. function to instruct computer to counter the most frequent weapon used by Human
    
    2. In case there is a conflict where two have the same frequency,
    
        The computer will choose the most optimum option as follows:
            Tie in R,P : Computer chooses P (Worst case: Human chooses P and its a draw; Best case: Computer wins)
            Tie in R,S : Computer chooses R (Worst case: Human chooses R and its a draw; Best case: Computer wins)
            Tie in S,P : Computer chooses S (Worst case: Human chooses S and its a draw; Best case: Computer wins)
   
   3. In case all three have the same frequency, computer will choose the weapon that defeats the most recent choice of Human.          

In [39]:
## Defining most frequent element in a list
def intersection(list1, list2): 
    """
    This function takes in two lists as input and gives out the common elements of the list
    """
    lst3 = [value for value in list1 if value in list2] 
    return lst3 

def most_frequent(test_list):
    """
    This function takes a list as an input and returns the most frequent element in the list.
    In case the frequency matches, it returns the 
    most recent one. 
    """
    R_freq = test_list.count('R')
    P_freq = test_list.count('P')
    S_freq = test_list.count('S')
    
    max_dict = {
        'R':R_freq,
        'P': P_freq,
        'S': S_freq}
    #print(max_dict)
    ls_max_choice = [k for k,v in max_dict.items() if v == max(max_dict.values())]
    #print(ls_max_choice)
    #print(len(ls_max_choice))
    ## Takes care of one frequent choice
    if len(ls_max_choice)==1:
        return str(ls_max_choice[0])
    elif len(ls_max_choice)==2:
        if len(intersection(['R','P'],ls_max_choice))==2:
            return 'R'
        elif len(intersection(['R','S'],ls_max_choice))==2:
            return 'S'
        else:
            return 'P'
    else:
        return test_list[-1]    

choice_comp:

    1. Function that calls the most frequent function and assigns a choice for the computer
    
    2. In case its the first iteration, computer is assigned a random weapon

In [40]:
def choice_comp(iterations, player_moves):
    """
        Function that calls the most frequent function and assigns a choice for the computer
        In case its the first iteration, computer is assigned a random weapon
    """
    if iterations == 0:
        comp_choice = int(np.random.randint(0,3,1))
        return choice_list[comp_choice]
    else:
        return winning_dictionary[most_frequent(player_moves)]

## Function to ensure that there is no bad input from the user

In [41]:
def Pchoice_func():
    """
        This will ensure that only good Inputs come from the Player
    """
    Pchoice = str(input("Please Choose Your Weapon (R:Rock; P:Paper ; S:Scissor ; Q: Quit) :")).upper()
    if Pchoice not in choice_list:
        print("ERROR!!! Choose a valid weapon")
        return Pchoice_func()
    else:
        return(Pchoice)


## Function to define winning criteria

In [42]:
def win_criteria(human,comp):
    """
        Deciding Criteria of Winning/Loosing/Draw
    """
    if human == comp : 
        return 0
    elif (human == 'P' and comp == 'R') or (human == 'R' and comp == 'S') or (human == 'S' and comp == 'P'):
        return 1
    elif human =='Q':
        return -99
    else: 
        return -1

## Main function of code

In [43]:
## Now lets start the main code to play the game
Pchoicelist = list([]) 
Cchoicelist = list([])
Pchoice = 'Null'
game_iterations = 0
games_result = list([])
while Pchoice != 'Q':
    # Computer Chooses first and keeps the choice a secret
    Cchoice = choice_comp(game_iterations,Pchoicelist)
    #Player chooses Second
    Pchoice = Pchoice_func()
    # print("Person :{} , Computer : {}".format(Pchoice , Cchoice))
    Cchoicelist.append(Cchoice) 
    res = win_criteria(Pchoice,Cchoice)
    if res == 0:
        print("It's a draw")
    game_iterations+=1
    games_result.append(res)
    Pchoicelist.append(Pchoice) 

print("You played {} games and won {} times".format(game_iterations-1 , games_result.count(1)))

It's a draw
It's a draw
It's a draw
It's a draw
You played 6 games and won 1 times


# Question-2: Voters in Florida (20 points)

The file _FloridaVoters.html_ contains a Web Table of republican and democratic voters in various counties in Florida.

Write code that reads in this fileas a standard text file and prints out the counties,
along with the numberof republican and democratic voters in those counties,sorted by the number of democratic voters.

## Importing the necessary libraries


In [44]:
import pandas as pd

# Mac Users:
# import regex as re

# Windows/Linux Users:
import re

# Reading the file

In [45]:
filename = "FloridaVoters.html"
with open(filename,"r") as f:  #Opening the file in read mode
    content = f.readlines()
    f.close()                  #Closing the file
print("Number of Lines:", len(content))

Number of Lines: 1396


# Extracting the table contents and cleaning them

In [46]:
list_rows = []
for row in content:
    cell = re.match('<td>|<th>',row)                       #Match the table headers and the table values
    if cell is not None:
        row1 = re.sub("<td>|<th>|</td>|</th>|\n|,","",row) #Removing the unnecessary tags
        list_rows.append(row1)                             #Appending all the rows into a list

# Creating List of Tuples

In [47]:
merged_list = [(list_rows[i], list_rows[i+1],list_rows[i+2],list_rows[i+3],list_rows[i+4],list_rows[i+5]) 
               for i in range(0, len(list_rows)-5,6)]

# Creating DataFrame and Sorting

In [48]:
df = pd.DataFrame(merged_list) #Converting to a dataframe
#Promoting Header
header_row = 0
df.columns = df.iloc[header_row]
df = df.drop(header_row)

df.reset_index(inplace = True)
df = df[['COUNTY','REPUBLICAN','DEMOCRAT']]

df['REPUBLICAN'] = df['REPUBLICAN'].astype('int')
df['DEMOCRAT'] = df['DEMOCRAT'].astype('int')
#Sorting the Dataframe
df = df.sort_values(by = ['DEMOCRAT'])

# Final Output

In [49]:
# Printing in the required format
def print_df(df):
   # This code delete index and columns names
    df_ = df.copy()
    df_.columns = ['', '', '']
    df_.index = ['' for _ in range(len(df))]
    print(df_)

In [50]:
print_df(df)

                                  
       LAFAYETTE     1373     2672
          GLADES     2190     3110
         LIBERTY      720     3372
           UNION     2752     3579
       GILCHRIST     5789     3652
..           ...      ...      ...
    HILLSBOROUGH   257436   314265
      PALM BEACH   245452   367236
      MIAMI-DADE   362161   539367
         BROWARD   249762   566185
           Total  4377713  4637026

[68 rows x 3 columns]


# Question-3: Near-duplicate detection (20 points)

Tweets on a subject are often nearly (but not exactly) duplicates of each other.

The  file **Santa.txt** contains a few tweets about Santa Claus, one tweet per line.

We will try to detect tweets that we have already seen before.

## Convert each tweet into a dictionary of phrases (6-points)

- Write a function called **moving_window** that takes as input a tweet (that is,a string), and converts it into a bunch of phrases.  Each phrase is three consecutive words in the tweet.
- For example, the tweet _"This is an awesome tweet, dude"_ consists of the phrases _"This is an"_,_"is an awesome"_,_"an awesome tweet,"_ and _"awesome tweet, dude"_.
- Return a dictionary whose keys are these phrases, and just set the corresponding values to 1.
- This dictionary contains all the unique 3-word phrases in the tweet.


In [51]:
def moving_window(tweet: str, size: int = 3):
    """
        Function that takes a tweet and phrase size as inputs.
        Returns Dictionary of consecutive word phrases of given "size"
    """
    wordList = tweet.lower().split()
    phraseList = dict()
    for index in range(len(wordList)-size+1):
        phrase = " ".join(wordList[index:index+size])
        if phrase not in phraseList.keys():
            phraseList[phrase] = 1
        else:
            phraseList[phrase] += 1
    return phraseList

moving_window("The quick Brown Fox jumped over the lazy dog")

{'the quick brown': 1,
 'quick brown fox': 1,
 'brown fox jumped': 1,
 'fox jumped over': 1,
 'jumped over the': 1,
 'over the lazy': 1,
 'the lazy dog': 1}

## Calculate Similarity between two tweets

To check if one tweet is a near-duplicate of another, we compute their cosine similarity:

_cosine(tweet1, tweet2)_ = matches/√(n1∗n2)

where matches is the number of 3-word phrases in common between the two tweets, and n1 and n2 are the number of phrases in the two tweets respectively.

- Write a function called **cosine** that takes as input two dictionaries.
- Each dictionary contains the 3-word phrases from one tweet.
- Return the cosine similarity between the phrases in the two dictionaries

In [52]:
import math
def Cosine(tweet1: dict, tweet2: dict):
    """
        Function to compare the cosine similarity between two tweets.
        Inputs: Tweet1 and Tweet2
        Output: Cosine Similarity Score
    """
    phraseSet1 = set(tweet1.keys())
    phraseSet2 = set(tweet2.keys())
    # print(phraseSet1)
    # print(phraseSet2)

    intersection = phraseSet1.intersection(phraseSet2)
    # print(intersection)

    return len(intersection)/math.sqrt(len(phraseSet1)*len(phraseSet2))

# sentence1 = "The quick Brown Fox jumped over the lazy dog"
# sentence2 = "the lazy dog is a man's best friend"
# Cosine(moving_window(sentence1),moving_window(sentence2))

## Read in tweets, and output near-duplicates (8 points)

- Read in the tweets in the file **Santa.txt**.
- For each tweet, figure out if it is a _near-duplicate_ of any of the previously-seen tweets.
- We say that the two tweets are _near-duplicates_ if their cosine similarity is greater than **0.5**.
- You should call the functions: _moving\_window_ and _cosine_ here

In [53]:
filename = "Santa.txt"
with open(filename,"r") as f:
    content = f.readlines()
    f.close()
print("Number of Tweets:", len(content))
# content

Number of Tweets: 5


In [54]:
counter = 0
for i in range(len(content)):
    for j in range(0,i):
        cosine_score = Cosine(moving_window(content[i]),moving_window(content[j]))
        if cosine_score >= 0.5:
            print("""
Tweet {i} is similar to Tweet {j}!\n\tScore: {score}
-------------
Tweet{i}: {tweet1}
Tweet{j}: {tweet2}""".format(score = cosine_score,
                        tweet1 = content[i].rstrip(),
                        tweet2 = content[j].rstrip(),
                        i = i+1, j = j+1))
            counter += 1
print("Total Matches: {}".format(counter))


Tweet 3 is similar to Tweet 1!
	Score: 0.8838834764831844
-------------
Tweet3: RT @BuyBookstore: SPECIAL SECRET HEARTS: A Child's Introduction to Dementia and Pink Curls - A Santa ... - http://t.co/UWCdc8FA9a http://t.Â…
Tweet1: SPECIAL SECRET HEARTS: A Child's Introduction to Dementia and Pink Curls - A Santa ... - http://t.co/UWCdc8FA9a http://t.co/meexKLGTKl

Tweet 4 is similar to Tweet 2!
	Score: 0.628970902033151
-------------
Tweet4: "RT @DrewFtDevonne_: Rt si te gusta Santa Claus Is Coming To Town #MTVHottest Justin Bieber"
Tweet2: "Santa Claus Is Coming To Town #MTVHottest Justin Bieber"

Tweet 5 is similar to Tweet 2!
	Score: 0.6837634587578276
-------------
Tweet5: "Rt si te gusta Santa Claus Is Coming To Town #MTVHottest Justin Bieber"
Tweet2: "Santa Claus Is Coming To Town #MTVHottest Justin Bieber"

Tweet 5 is similar to Tweet 4!
	Score: 0.8362420100070908
-------------
Tweet5: "Rt si te gusta Santa Claus Is Coming To Town #MTVHottest Justin Bieber"
Tweet4: "RT @DrewFtDe

# Question-4: The Google of Quotes (40 points)

 The file “quotes.txt” contains pairs of lines, with the first line being a quoteand  the  following  line  being  the  person  who  said  it.   We  want  to  build  asearch engine that, given a word or words, finds the best matching quotes.

In [55]:
# Mac Users:
# import regex as re

# Windows/Linux Users:
import re

In [56]:
filename = "quotes.txt"
with open(filename,"r") as f:
    content = f.readlines()
    f.close()
print("Number of Lines:", len(content))

Number of Lines: 1772


## Build a list of full quotes (5 points)

Read in the file, and create a list of full quotes of the form “quote - speaker”.

For example,“The heart has its reasons, of which the mind knows nothing. - Blaise Pascal”.

In [57]:
content = [" - ".join([content[i].rstrip(),content[i+1].rstrip()]) for i in range(0,len(content),2)]
# print(len(content))
content[0:5]

['How we spend our days is, of course, how we spend our lives. - Annie Dillard',
 'Two roads diverged in a wood, and I...I took the one less traveled by, and that has made all the difference. - Robert Frost',
 'What is happiness? The feeling that power is growing, that resistance is overcome. - Friedrich Nietzsche',
 'A great deal of intelligence can be invested in ignorance when the need for illusion is deep. - Saul Bellow',
 'Those who are preoccupied with `making a statement` usually don`t have any statements worth making. - Thomas Sowell']

## Words from full quotes (5 points)

- Write  a  function  that  takes a full  quote  as  argument  and  outputs  a  list  of  the  words  in  the  it.
- The words should all be lower-case, and should contain only characters, digits, or underscore


In [58]:
def WordSplit(quote: str):
    """
        Split quotes into words.
    """
    # Regex to clear string.
    # String can contain only Characters, Digits or underscore
    quote = re.sub('[^A-Za-z0-9_ ]', '', quote)
    return quote.lower().replace(" - "," ").replace("."," ").split()

WordSplit(content[5])

['women',
 'need',
 'a',
 'reason',
 'to',
 'have',
 'sex',
 'men',
 'just',
 'need',
 'a',
 'place',
 'billy',
 'crystal']

## Build the postings-list dictionary (6 points)

A postings-list is a dictionary  whose  keys  are  full  quotes,  and  whose  values  are  themselves dictionaries  with  key  being  a  word,  and  value  being  the  number  of  times the  word  occurs  in  the  full  quote.

So, for the **key** _"The heart has its reasons, of which the mind knows nothing. - Blaise Pascal"_,

the value will itself be the following  dictionary:

```Python
{"the":2, "heart":1, "has":1, "its":1, "reasons":1, "of":1, "which":1, "mind":1, "knows":1, "nothing":1, "blaise":1, "pascal":1}
```


In [59]:
def TfDict(words: tuple):
    """
        Function that takes tuple of words as input
        Returns PostingsList Dictionary as output
    """
    tfDict = dict()
    for word in words:
        if word not in tfDict.keys():
            tfDict[word] = 1
        else:
            tfDict[word] += 1
    return tfDict

# list(map(TfDict,map(WordSplit, content)))[0]

In [60]:
postingsList = dict()
for quote in content:
    if quote not in postingsList.keys():
        postingsList[quote] = TfDict(tuple(WordSplit(quote)))
    else:
        print("Duplicate Quote:\n{quote}".format(quote = quote))

postingsList['How we spend our days is, of course, how we spend our lives. - Annie Dillard']

{'how': 2,
 'we': 2,
 'spend': 2,
 'our': 2,
 'days': 1,
 'is': 1,
 'of': 1,
 'course': 1,
 'lives': 1,
 'annie': 1,
 'dillard': 1}

# Build the reverse postings-list dictionary (6 points)

A reverse postings-list is a dictionary whose keys are the words, and the values are themselves dictionaries with the key being a full quote, and the value being the number of times the word appeared in the full quote.

So, for the key _"entertainer"_, the value is the dictionary

```Python
{"An actor is at most a poet and at least an entertainer. - Marlon Brando":  1} 
#only this quote contains the word “entertainer”
```


In [61]:
# Create Set of all words
allWords = set(" ".join(content).lower().replace(" - "," ").replace("."," ").split())
# print(len(allWords))

In [62]:
def IdfDict(word: str):
    """
        Function that takes a words as input
        Returns Reverse-PostingsList Dictionary as output
    """
    idfDict = dict()
    for quote in content:
        if word in postingsList[quote].keys():
            idfDict[quote] = postingsList[quote][word]
    return idfDict

# IdfDict("spend")

In [63]:
revPostingsList = dict()
for word in allWords:
    if word not in revPostingsList.keys():
        revPostingsList[word] = IdfDict(word)
    else:
        print("Duplicate Word:\n{word}".format(word = word))

revPostingsList["book"]

{'The Bible is not my book, and Christianity is not my religion. I could never give assent to the long, complicated statements of Christian dogma. - Abraham Lincoln': 1,
 'There is a great deal of difference between an eager man who wants to read a book and the tired man who wants a book to read. - GK Chesterton': 2,
 'The covers of this book are too far apart. - Ambrose Bierce': 1,
 'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book. - Marcel Proust': 1,
 'A truly great book should be read in youth, again in maturity and once more in old age, as a fine building should be seen by morning light, at noon and by moonlight. - Robertson Davies': 1,
 'Don`t join the book burners. Do not think you are going to conceal thoughts by concealing evidence that they ever existed. - Dwight Eisenhower': 1}

# Write a TF-IDF function (8 points)

To measure how much a full quote is about a particular word, one typically uses the TF-IDF measure

## TF

- TF  stands  for  **term  frequency**
- The  term  frequency  of  a  word _w_ in a  full  quote _q_ is  the  **number  of  times** _w_ occurs  in q,  divided  by  the maximum number of times any word occurs in _q_

## IDF

- IDF stands for **inverse document frequency**
- The IDF of a word _w_ is the logarithm of the ratio of the total number _N_ of full quotes to the number of full quotes in that contain the word _w_

TF-IDF of a word _w_ for a full quote _q_ is just the **product** of the **TF** and **IDF**.

So, for the word _entertainer_ in the Marlon Brando quote of part (d):

- The  TF  is  0.5  (it  occurs  once,  while  the  most  frequent  word  in  that quote is “at”, which occurs twice, so the TF ratio is 0.5)
- The IDF is log (886/1),  since there are 886 documents and the word“entertainer ”occurs in only one full quote.

Write a function to compute the TF-IDF of any word in any full quote, using the postings and reverse-postings.

Hint:Do import math and use _math.log()_ to get logarithms.


In [64]:
import operator
def MaxWordFreqinQuote(quote: str):
    """
        Function to return the Frequency of the most frequent word in a quote
    """
    return max(postingsList[quote].items(), key=operator.itemgetter(1))[1]

In [65]:
import math
def TfIdf(word: str, quote: str):
    """
        Function to return the Tf-Idf score for a given word in a given quote
        Inputs: Word and Quote
        Outputs: Tf-Idf Score
    """
    tf, idf = 0,0

    # Check if Quote Exists and Word exists in Quote. Else, return 0 (b/c TF = 0)
    if (quote in postingsList.keys())and (word in postingsList[quote].keys()):
        tf = postingsList[quote][word]/MaxWordFreqinQuote(quote)
    
    # Check if Word exists in revPostingList. Else, return 0 (b/c IDF = 0)
    if word in revPostingsList.keys():
        idf = math.log(len(content) / len(revPostingsList[word].keys()))
    else:
        return "Word Does not exist"
    # print("TF: {tfscore};\nIDF: log({N}/{x}) = {idfscore}".format(tfscore = tf,
    #                                                             idfscore = idf,
    #                                                             N = len(content),
    #                                                             x = len(revPostingsList[word].keys())))
    return tf*idf

In [66]:
TfIdf("entertainer","An actor is at most a poet and at least an entertainer. - Marlon Brando")

3.3933584753025405

## Quote  search  using  a  single  word  (5  points)

Write  a  function that takes a word as argument, and returns a dictionary whose keys are full quotes containing that word, and whose values are the TF-IDF score of that word for that full quote.


In [67]:
def SingleWordScore(word: str):
    """
        Input: Word 
        Returns: A dictionary:   KEYS -> full quotes containing that word
                                VALUE -> TF-IDF score of the word for that full quote
    """
    score = dict()
    if word in revPostingsList.keys():
        for quote in revPostingsList[word]:
            score[quote] = TfIdf(word = word, quote= quote)
    return score

SingleWordScore("spent")

{'I feel impelled to speak today in a language that in a sense is new-one which I, who have spent so much of my life in the military profession, would have preferred never to use. That new language is the language of atomic warfare. - Dwight Eisenhower': 2.0311899233483786,
 'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book. - Marcel Proust': 3.046784885022568}

In [68]:
SingleWordScore("book")

{'The Bible is not my book, and Christianity is not my religion. I could never give assent to the long, complicated statements of Christian dogma. - Abraham Lincoln': 2.497478740688513,
 'There is a great deal of difference between an eager man who wants to read a book and the tired man who wants a book to read. - GK Chesterton': 3.3299716542513504,
 'The covers of this book are too far apart. - Ambrose Bierce': 4.994957481377026,
 'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book. - Marcel Proust': 2.497478740688513,
 'A truly great book should be read in youth, again in maturity and once more in old age, as a fine building should be seen by morning light, at noon and by moonlight. - Robertson Davies': 1.6649858271256752,
 'Don`t join the book burners. Do not think you are going to conceal thoughts by concealing evidence that they ever existed. - Dwight Eisenhower': 4.994957481377026}

## Quote search using multiple words (5 points)

Write a function that takes a list of words as argument, and returns a dictionary whose keys are full quotes containing one or more of the words in that list, and whose values are the **sum of TF-IDF scores** of the words in that list for that full-quote.


In [69]:
def MultiWordScore(wordList: list):
    """
        Input: List of Words
        Returns: A dictionary:   KEYS -> full quotes containing that word(s)
                                VALUE -> sum of TF-IDF score of the word(s) for that full quote
    """
    score = dict()
    for word in wordList:
        if word in revPostingsList.keys():
            for quote in revPostingsList[word]:
                if quote in score.keys():
                    score[quote] += TfIdf(word = word, quote= quote)
                else:
                    score[quote] = TfIdf(word = word, quote= quote)
    return score

MultiWordScore(["spent","book","entertainer"])

{'I feel impelled to speak today in a language that in a sense is new-one which I, who have spent so much of my life in the military profession, would have preferred never to use. That new language is the language of atomic warfare. - Dwight Eisenhower': 2.0311899233483786,
 'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book. - Marcel Proust': 5.54426362571108,
 'The Bible is not my book, and Christianity is not my religion. I could never give assent to the long, complicated statements of Christian dogma. - Abraham Lincoln': 2.497478740688513,
 'There is a great deal of difference between an eager man who wants to read a book and the tired man who wants a book to read. - GK Chesterton': 3.3299716542513504,
 'The covers of this book are too far apart. - Ambrose Bierce': 4.994957481377026,
 'A truly great book should be read in youth, again in maturity and once more in old age, as a fine building should be seen by morning light, at noon a

<br><p>
**TF-IDF** score of _**spent**_ in _'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book - Marcel Proust'_ : **3.0467**
<br><p>
**TF-IDF** score of _**book**_ in _'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book - Marcel Proust'_ : **2.4974**
<br><p>
**TF-IDF** score of _**spent, book**_ in _'There are perhaps no days of our childhood we lived so fully as those we spent with a favorite book - Marcel Proust'_ : **5.54426**